applications/RomApplication/custom_strategies/lspg_rom_builder_and_solver.h

```diff
@@ -0,0 +1,587 @@
+    ///@{
+
+    /// Turn back information as a string.
+    virtual std::string Info() const override
+    {
+        return "LSPGROMBuilderAndSolver";
+    }
+
+    /// Print information about this object.
+    virtual void PrintInfo(std::ostream &rOStream) const override
+    {
+        rOStream << Info();
+    }
+
+    /// Print object's data.GetNumberOfROMModes()
+    virtual void PrintData(std::ostream &rOStream) const override
+    {
+        rOStream << Info();
+    }
+
+    ///@}
+    ///@name Friends
+    ///@{
+
+
+    ///@}
+protected:
+
+    SizeType mNodalDofs;
+    bool mTrainPetrovGalerkinFlag = false;
+
+    ///@}
+    ///@name Protected static member variables
+    ///@{
+
+
+    ///@}
+    ///@name Protected member variables
+    ///@{
+
+    ///@}
+    ///@name Protected operators
+    ///@{
+
+
+    ///@}
+    ///@name Protected operations
+    ///@{
+
+    void AssignSettings(const Parameters ThisParameters) override
+    {
+        BaseType::AssignSettings(ThisParameters);
+
+        // // Set member variables
+        mTrainPetrovGalerkinFlag = ThisParameters["train_petrov_galerkin"].GetBool();
+    }
+
+    /**
+    * Thread Local Storage containing dynamically allocated structures to avoid reallocating each iteration.
+    */
+    struct AssemblyTLS
+    {
+        Matrix phiE = {};                // Elemental Phi
+        LocalSystemMatrixType lhs = {};  // Elemental LHS
+        EquationIdVectorType eq_id = {}; // Elemental equation ID vector
+        DofsVectorType dofs = {};        // Elemental dof vector
+        RomSystemMatrixType romA;        // reduced LHS
+        RomSystemVectorType romB;        // reduced RHS
+    };
+
+    /**
+     * Resizes a Matrix if it's not the right size
+     */
+    template<typename TMatrix>
+    static void ResizeIfNeeded(TMatrix& mat, const SizeType rows, const SizeType cols)
+    {
+        if(mat.size1() != rows || mat.size2() != cols) {
+            mat.resize(rows, cols, false);
+        }
+    };
+
+    /**
+     * Builds the reduced system of equations on rank 0 
+     */
+    void BuildROM(
+        typename TSchemeType::Pointer pScheme,
+        ModelPart &rModelPart,
+        LSPGSystemMatrixType &rA,
+        LSPGSystemVectorType &rb) override
+    {
+        KRATOS_TRY
+        // Define a dense matrix to hold the reduced problem
+        rA = ZeroMatrix(BaseType::GetEquationSystemSize(), this->GetNumberOfROMModes());
+        rb = ZeroVector(BaseType::GetEquationSystemSize());
+
+        // Build the system matrix by looping over elements and conditions and assembling to A
+        KRATOS_ERROR_IF(!pScheme) << "No scheme provided!" << std::endl;
+
+        // Get ProcessInfo from main model part
+        const auto& r_current_process_info = rModelPart.GetProcessInfo();
+
+
+        // Assemble all entities
+        const auto assembling_timer = BuiltinTimer();
+
+        AssemblyTLS assembly_tls_container;
+
+        auto& elements = rModelPart.Elements();

```

Comment:
```suggestion
        const auto& r_elements = rModelPart.Elements();

```
Note the `const`. I think this is a good candidate.